Topic fix server options numAudioBusChannels

telephon opened the pull request:
This makes ServerOptions numAudioBusChannels calculations more perspicuous and adds a unit test for them.

telephon commented:
This is a first step at solving #2758.

brianlheim commented on SCClassLibrary/Common/Control/Server.sc:
```diff
@@ -38,8 +40,6 @@ ServerOptions {
 	var <>threads = nil; // for supernova
 	var <>useSystemClock = false;  // for supernova
 
-	var <numPrivateAudioBusChannels=112;
```
Comment at start of file says "order of variables is important here. Only add new instance variables to the end"... is this something to worry about?

telephon replied:
Yes, I worried, too. I've had a look and found no reference. Actually I am not sure if that still applies at all - I don't see where it should.

brianlheim replied:
Intrinsic classes have their instance/class variables hardcoded on the C++ side. That's definitely true of things like Finalizer and the classes in Kernel. It doesn't seem to be the case here.

telephon replied:
yes exactly.

brianlheim replied:
Actually, I think I've found some hard linking that occurs here: 

https://github.com/supercollider/supercollider/blob/master/lang/LangPrimSource/OSCData.cpp#L983-L1072

But, everything past `inDevice`/`outDevice` should be fine.

telephon replied:
ah good. I'll add a note.

brianlheim commented on testsuite/classlibrary/TestServerOptions.sc:
```diff
@@ -0,0 +1,30 @@
+		testAudio.value;
+		options.numInputBusChannels = rrand(4, 8);
+		testAudio.value;
+		options.numOutputBusChannels = rrand(4, 8);
```
These tests would be more useful if they tested ranges of values (or maxima and minima) rather than picking randomly. Are these the full allowed ranges for these values? (Is that enforced?)

telephon replied:
We might need to add a check in the class first that catches unsane values. I'll also do better tests for bus allocation. We need to take into account `clientID`, too.

brianlheim replied:
Thank you!

jamshark70 commented:
> Intrinsic classes have their instance/class variables hardcoded on the C++ side. That's definitely true of things like Finalizer and the classes in Kernel. It doesn't seem to be the case here.

I believe they are intrinsic to support the internal server. Be careful not to break that.

brianlheim commented:
> I believe they are intrinsic to support the internal server. Be careful not to break that.

Absolutely. I think this needs further investigation before we can merge. I posted a code segment above that relates, but it may not be the only part of the underlying implementation that directly pulls from hard-coded slot positions.

brianlheim commented on SCClassLibrary/Common/Control/Server.sc:
```diff
@@ -1,10 +1,12 @@
 ServerOptions {
-	// order of variables is important here. Only add new instance variables to the end.
+
+	// order of variables is important here. You may add new instance variables after outDevice.
```
As I noted below, this is only my interpretation at a quick glance at one code file. I'd really prefer it if this was verified more thoroughly so we don't commit misinformation.

telephon replied:
ok, I'll revert this.

jamshark70 commented:
Writing mainly because my opinion was solicited under #2758. I'm afraid I'm not likely to have time to unpack everything that's changing here. It *appears* that it's trying not to change the public interface, but in the absence of a spec, it's quite confusing.

So:

1. Bus configuration in ServerOptions, and usage in `newAllocators`, is not obvious and there is no written spec. While I can go along with others' suggestion elsewhere that it's too onerous to require a spec for everything, this is a particular case where we got into trouble because nobody knows exactly how it's really supposed to work. Maybe it would be better to write that down somewhere (wiki?).

2. If this is not an API change, I'm OK with it. (By the way, does it change the meaning of the variable names?)

telephon commented:
Sure, it is an API change (though I doubt a bit if this term will help us too much in the future). The thing is that the spec never really existed, and this is how we usually do things (without explicit double specification). 

Looking at the whole code now, I am inclined to remove all the checks in the ServerOptions again (make them just setters) and only check when the server tries to boot, and fail with an informative message if the user has set the wrong combination.

telephon commented:
Check the alternative here: #2774

telephon commented:
@adcxyz what do you think: are you addressing the issues for which this PR is a fix? If not, I'd reopen it and see if we can merge it. I'd like to resolve this old thing …
